Singularity/Singularity/Platform/IPlatformAction.cs

```diff
@@ -0,0 +1,140 @@
+
+        /// <summary>
+        /// Gets the JobType required for this PlatformAction. Can be several.
+        /// (e.g. Factory needs Carrying units and Producing units,
+        /// and every Blueprint needs Building units, ...)
+        /// </summary>
+        /// <value>The units required.</value>
+        List<JobType> UnitsRequired { get; }
+
+        /// <summary>
+        /// Gets the assigned units and their respective JobTypes.
+        /// </summary>
+        /// <value>The assigned units.</value>
+        Dictionary<GeneralUnit, JobType> AssignedUnits { get; }
+
+        PlatformBlank Platform { get; }
+    }
+
+
+
+
+    public abstract class APlatformAction : IPlatformAction
+    {
+        protected readonly Dictionary<GeneralUnit, JobType> mAssignedUnits = new Dictionary<GeneralUnit, JobType>();
+        protected readonly PlatformBlank mPlatform;
+
+        protected APlatformAction(PlatformBlank platform)
+        {
+             mPlatform = platform;
+        }
+
+
+        public PlatformActionState State { get; private set; } = PlatformActionState.Active;
+
+        public abstract List<JobType> UnitsRequired { get; }
+
+        PlatformBlank IPlatformAction.Platform => mPlatform;
+        Dictionary<GeneralUnit, JobType> IPlatformAction.AssignedUnits => mAssignedUnits;
+
+        /// <summary>
+        /// Assigns the unit to this PlatformAction and to this platform.
+        /// </summary>
+        /// <param name="unit">Unit.</param>
+        /// <param name="job">Job.</param>
+        void IPlatformAction.AssignUnit(GeneralUnit unit, JobType job)
+        {
+            mAssignedUnits.Add(unit, job);
+        }
+
+        public abstract void Execute();
+
+        /// <summary>
+        /// Gets the required resources of the PlatformAction
+        /// (for finishing this action, or for producing the next resource etc)
+        /// </summary>
+        /// <returns>The required resources.</returns>
+        Dictionary<EResourceType, int> IPlatformAction.GetRequiredResources()
+        {
+            return new Dictionary<EResourceType, int>();
+        }
+
+        void IPlatformAction.UiToggleState()
+        {
+            switch (State)
+            {
+                case PlatformActionState.Available:
+                    State = PlatformActionState.Deactivated;
+                    break;
+                case PlatformActionState.Deactivated:
+                    State = PlatformActionState.Available;
+                    break;
+                default:
+                    throw new AccessViolationException("Someone/Something acccessed the state!!");
+            }
+        }
+
+        public void UnAssignUnits(int amount, JobType job)
+        {
+            foreach (var unit in mAssignedUnits.Keys)
+            {
+                if (unit.Job != job || amount <= 0) continue;
+                mAssignedUnits.Remove(unit);
+                amount -= 1;
+            }
+        }
+    }
+}
```

Comment:
Is it ok to have the abstract platform inside of the interface? I'm not sure here so if it's fine then it's fine

Reply:
Yes this is alright, since it'd be redundant code otherwise. It's actually only an abstract class, more or less with 'default'-implementations for certain parts of PlatformActions.